File: VerifyAlien.java
// TC:O(nk)   ...k= avg length of words
// SC:O(1)

class Solution {
    HashMap<Character, Integer> map;
    public boolean isAlienSorted(String[] words, String order) {
        
        map = new HashMap();
        for(int i=0;i<order.length();i++){
            map.put(order.charAt(i),i);
        }
        
        int n= words.length;
        for(int i=0;i<n-1;i++){
            String first= words[i];
            String second= words[i+1];
            if(isNotSorted(first,second)) return false;
        }
        return true;
    }
    
    private boolean isNotSorted(String first, String second){
        for(int i=0; i< first.length() && i< second.length(); i++){
            char fChar= first.charAt(i);
            char sChar= second.charAt(i);
            
            if(fChar!=sChar){
                return map.get(fChar)>map.get(sChar);
            }
        }
        
        if(first.length()> second.length()) return true;
        return false;
    }
}

File: verifyAlien.py
'''
time complexity: O(nk)
k is avg len of words
space complexity: O(1)
'''
class Solution:
    def isAlienSorted(self, words: List[str], order: str) -> bool:
        self.count = {}
        c = 0
        for o in order:
            self.count[o] = c
            c+=1
        for i in range(len(words)-1):
            fw = words[i]
            sw = words[i+1]
            if(self.isNotSmaller(fw,sw)): return False
        return True
    
    def isNotSmaller(self,fw,sw):
        i = 0
        j = 0
        while(i<len(fw) and j<len(sw)):
            if(fw[i]!=sw[j]):
                if(self.count[fw[i]] > self.count[sw[j]]): return True
                return False
            i+=1
            j+=1
        if(len(fw) > len(sw)): return True
        return False
            